sycl/doc/SYCLNativeCPU.md

```diff
@@ -0,0 +1,170 @@
+clang++ -L<sycl-lib-path> -lsycl <device-o> <host-o> -o <output>
+```
+In order to execute kernels compiled for `native-cpu`, we provide a PI Plugin. The plugin needs to be enabled when configuring DPC++ (e.g. `python buildbot/configure.py --enable-plugin native_cpu`) and needs to be selected at runtime by setting the environment variable `ONEAPI_DEVICE_SELECTOR=native_cpu:cpu`. 
+
+# Supported features and current limitations
+
+The SYCL Native CPU flow is still WIP, not optimized and several core SYCL features are currently unsupported. Currently `barrier` and all the math builtins are not supported, and attempting to use those will most likely fail with an `undefined reference` error at link time. Examples of supported applications can be found in the [runtime tests](sycl/test/native_cpu).
+
+
+To execute the `e2e` tests on the Native CPU, configure the test suite with:
+
+```bash
+# make sure that DPC++ is in your $PATH and your environment is configured for DPC++
+
+cd sycl/test-e2e
+cmake \
+  -G Ninja \
+  -B build -S . \
+ -DCMAKE_CXX_COMPILER=clang++ \
+ -DSYCL_TEST_E2E_TARGETS="native_cpu:cpu" 
+
+```
+
+Note that a number of `e2e` tests are currently still failing.
+
+# Running example
+
+The following section gives a brief overview of how a simple SYCL application is compiled for the Native CPU target. Consider the following SYCL sample, which performs vector addition using USM:
+
+```c++
+  cl::sycl::queue deviceQueue;
+  cl::sycl::range<1> numOfItems{N};
+  auto a_ptr = sycl::malloc_device<int>(N, deviceQueue);
+  auto b_ptr = sycl::malloc_device<int>(N, deviceQueue);
+  auto c_ptr = sycl::malloc_device<int>(N, deviceQueue);
+
+  // copy mem to device, omitted
+  deviceQueue
+      .submit([&](cl::sycl::handler &cgh) {
+        auto kern = [=](cl::sycl::id<1> wiID) {
+          c_ptr[wiID] = a_ptr[wiID] + b_ptr[wiID];
+        };
+        cgh.parallel_for<class Sample>(numOfItems, kern);
+      })
+      .wait();
+  deviceQueue.memcpy(C.data(), c_ptr, N * sizeof(int));
+
+```
+
+The extracted device code produces the following LLVM-IR:
+
+```llvm
+define weak_odr dso_local spir_kernel void @_Z6Sample(ptr noundef align 4 %_arg_c_ptr, ptr noundef align 4 %_arg_a_ptr, ptr noundef align 4 %_arg_b_ptr) local_unnamed_addr #1 comdat !srcloc !74 !kernel_arg_buffer_location !75 !kernel_arg_type !76 !sycl_fixed_targets !49 !sycl_kernel_omit_args !77 {
+entry:
+  %0 = load i64, ptr @__spirv_BuiltInGlobalInvocationId, align 32, !noalias !78
+  %arrayidx.i = getelementptr inbounds i32, ptr %_arg_a_ptr, i64 %0
+  %1 = load i32, ptr %arrayidx.i, align 4, !tbaa !72
+  %arrayidx4.i = getelementptr inbounds i32, ptr %_arg_b_ptr, i64 %0
+  %2 = load i32, ptr %arrayidx4.i, align 4, !tbaa !72
+  %add.i = add nsw i32 %1, %2
+  %cmp.i8.i = icmp ult i64 %0, 2147483648
+  tail call void @llvm.assume(i1 %cmp.i8.i)
+  %arrayidx6.i = getelementptr inbounds i32, ptr %_arg_c_ptr, i64 %0
+  store i32 %add.i, ptr %arrayidx6.i, align 4, !tbaa !72
+  ret void
+}
```

Comment:
addrspace is not shown for kernel pointer arguments. Is address space not generated for them?

Reply:
Thanks for your comment, currently we set the target triple to the same as the host triple, which leads CodeGen using the X86AddrSpaceMap, so addrspaces are not emitted. I believe that the only information about addrspace that we need is whether or not an argument requires local memory, so we will address this when adding support for local memory (I'll add workgroup local memory in the ongoing work section of our docs)

Reply:
it should be also feasible to use spir/spir64 triple for device code generation and then materialize ir to the host triple. What is the benefit of using host triple directly in device code generation?

Reply:
Yeah that's another option, we haven't tried it yet - we may try it if the lack of addrspaces turns out to be a problem for us, but at this stage it seemed more "natural" to use the same host triple during device code generation, and personally I'm not sure about issues that may arise when lowering/linking if the choose to materialize the target triple in a later stage.